File: MEKF/MEKF_py/update.py
import numpy as np
import math
from numpy import linalg as LA
def quatMult(q1,q2):
    qn = np.zeros(4)
    qn[0] = q1[0]*q2[0]-np.matmul(q1[1:4].T,q2[1:4])
    
    qn_temp = q1[0]*q2[1:4]+q2[0]*q1[1:4]+np.reshape(np.cross(q1[1:4].T,q2[1:4].T),(3,1))
    
    qn[1:4] = np.reshape(qn_temp,3)
    qn = qn / LA.norm(qn)
    # propagate dynamics forward
    return np.reshape(qn,(4,1))

def update(L, z, xn, Pn, V, C):
    
    dx = L@z
    dphi = dx[0:3]
    dbeta = dx[3:6]

    temp = math.sqrt(1-np.reshape(dphi.T@dphi,1))
    temp_q = np.vstack((temp,dphi))
    xn = np.reshape(xn,(7,1))
    q = quatMult(xn[0:4],temp_q)

    b = xn[4:7] + dbeta
    b = np.reshape(b,(3,1))
    
    x = np.vstack((q,b))
    P = (np.identity(6)-L*C)*Pn*np.transpose(np.identity(6)-L*C)+L*V*np.transpose(L)
    
    return x, P

File: MEKF/MEKF_C/update.cpp
#include <iostream>
#include "eigen-git-mirror/Eigen/Dense"

using Eigen::MatrixXd;

MatrixXd quatmult(MatrixXd q1, MatrixXd q2);
MatrixXd cross(MatrixXd a, MatrixXd b);

MatrixXd quatmult(MatrixXd q1, MatrixXd q2){
    MatrixXd qn_temp(4,1);
    qn_temp(0,0) = q1(0,0)*q2(0,0)-q1(1:4,0).transpose()*q2(1:4,0);
    qn_temp(1:4,0) = q1(0,0)*q2(1:4,0)+q2(0,0)*q1(1:4,0)+cross(q1(1:4,0),q2(1:4,0));
    MatrixXd qn(4,1);
    qn = qn_temp / sqrt(qn_temp(0,0)^2+qn_temp(1,0)^2+qn_temp(2,0)^2+qn_temp(3,0)^2);
    return qn;
}

MatrixXd cross(MatrixXd a, MatrixXd b){
    MatrixXd c(3,1);
    c(0,0) = a(1,0)*b(2,0)-a(2,0)*b(1,0);
    c(1,0) = -a(0,0)*b(2,0)+a(2,0)*b(0,0);
    c(2,0) = a(0,0)*b(1,0)-a(1,0)*b(0,0);
    return c;
}

MatrixXd update(MatrixXd L, MatrixXd z, MatrixXd xn, MatrixXd Pn, MatrixXd V, MatrixXd C){
    MatrixXd dx(6,1) = L * z;
    MatrixXd dphi(3,1) = dx(0:2,0);
    MatrixXd dbeta(3,1) = dx(3:5,0);
    double temp = sqrt(1-dphi.transpose()*dphi);
    MatrixXd temp_q(4,1);
    temp_q(0,0) = temp;
    temp_q(1:4,0) = dphi;
    MatrixXd q(4,1);
    q = quatmult(xn(0:4,0),temp_q);
    MatrixXd b(3,1);
    b = xn(4:7,0) + dbeta;
    MatrixXd x(7,1);
    x(0:4,0) = q;
    x(4:,0) = b;
    MatrixXd P(6,6);
    P = (identity(6)-L*C)*Pn*(identity(6)-L*C).transpose() + L*V*L.transpose();
    return x, P;
}
